File: Java/src/0101 ~ 0200/0117 - Populating Next Right Pointers in Each Node II.java
// 链接：https://leetcode.com/problems/populating-next-right-pointers-in-each-node-ii/
// 题意：给定一个二叉树，结点的定义如下：
//      struct Node {
//        int val;
//        Node *left;
//        Node *right;
//        Node *next;
//      }
//      初始情况下每个结点的 next 都为空，
//      现在需要处理成每个结点的 next 为当前一层的下一个结点？
//      （使用 O(1) 的额外空间，本题递归栈不算作额外空间）

// 输入： root = [1,2,3,4,5,null,7]
// 输出： [1,#,2,3,#,4,5,7,#]
// 解释： 输出是按照每一层通过 next 进行的遍历，每一层通过 # 分隔

// 思路：循环
//
//		思路同 0117 一样，不过需要加一点变化适配当前题目
//
//      1. 第 1 层的 next 不需要处理即符合题意
//      2. 假设第 i 层的 next 已经全部处理完
//         设第 i 层对应的最左边的结点分别为 leftMost
//         每一层的初始状态如下： cur = leftMost ，并记录第 i + 1 层在链上第最后一个结点 last
//         则可以通过以下方式更新第 i + 1 层的 next （注意：还需要记录下一层最左边的结点）
//         (1) cur 的左子结点存在，则将其挂在链上，可以进行更新：
//              last.next = cur.left
//              last = cur.left
//         (1) cur 的右子结点存在，则将其挂在链上，可以进行更新：
//              last.next = cur.right
//              last = cur.right
//
//      时间复杂度： O(n)
//      空间复杂度： O(1)

/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/
class Solution {
    public Node connect(Node root) {
        Node leftMost = root;
        // 若 leftMost 存在，则可以继续处理下一层的 next
        while (leftMost != null) {
            Node cur = leftMost;
            // 用一个空结点表示第 i + 1 层链上第最后一个结点，方便统一逻辑
            Node last = new Node();
            // 下一层最左边的结点
            Node nextLeftMost = null;

            // 若 cur 存在，则可更新 i + 1 层
            while (cur != null) {
                // 如果左子结点存在，则将其挂在链上
                if (cur.left != null) {
                    // 如果还未找到下一层最左边的结点，则 cur.left 就是
                    if (nextLeftMost == null) {
                        nextLeftMost = cur.left;
                    }
                    last.next = cur.left;
                    last = last.next;
                }
                // 如果左子结点存在，则将其挂在链上
                if (cur.right != null) {
                    // 如果还未找到下一层最左边的结点，则 cur.left 就是
                    if (nextLeftMost == null) {
                        nextLeftMost = cur.right;
                    }
                    last.next = cur.right;
                    last = last.next;
                }
                // 处理下一个结点
                cur = cur.next;
            }
            // 继续处理下一层
            leftMost = nextLeftMost;
        }
        return root;
    }
}

File: Python3/src/0101 ~ 0200/0117 - Populating Next Right Pointers in Each Node II.py
# 链接：https:#leetcode.com/problems/populating-next-right-pointers-in-each-node-ii/
# 题意：给定一棵二叉树，每个结点 cur 有 val, left, right, next 四个字段，
#      其中 cur.next 指向当前层 cur 右侧的结点，
#      如果 cur 右侧没有结点，则设置为 NULL 。
#
#      初始时所有结点的 next 都是 NULL ，我们需要设置正确的 next 。


# 数据限制：
#  树的结点数在 [0, 6000] 内
#  -100 <= Node.val <= 100


# 输入： root = [1,2,3,4,5,null,7]
# 输出： [1,#,2,3,#,4,5,7,#]
# 解释： 1            1    
#      / \          / \
#     2   3   =>   2 → 3
#    / \   \      / \   \
#   4   5   7    4 → 5 → 7
#
#      序列化的结果是按照每层的 next 顺序处理的，
#      每一层的末尾用 '#' 标示。

# 输入： root = []
# 输出： []


# 思路： 迭代
#
#      假设第 i 层的 next 已经全部处理完，
#      且第 i 层对应的最左侧结点为 left_most 。
#
#      则我们可以通过遍历 left_most 基于 next 的链表，
#      将第 i + 1 层的所有结点的 next 处理完成。
#
#      第 1 层的 next 不需要处理即符合题意，其 left_most = root 。
#
#      处理第 i 层的 left_most 时，
#      我们使用 cur 遍历当前层的结点，初始化 cur = left_most 。
#
#      同时我们维护第 i + 1 层的链表结点 next_head_pre 和 next_last ，
#      其中 next_head_pre 是第 i + 1 层链表的哨兵结点，方便后续处理，
#      next_last 是第 i + 1 层链表的最后一个结点，方便使用尾插法插入新结点。
#
#      遍历过程中，判断 cur 的左右子结点：
#         1. cur 的左子结点存在，则将其挂在链上，可以进行更新：
#              last.next = cur.left
#              last = cur.left
#         2. cur 的右子结点存在，则将其挂在链上，可以进行更新：
#              last.next = cur.right
#              last = cur.right
#
#      然后更新 cur = cur.next ，直至 cur 为空。
#
#      此时，我们将 next_head_pre.next 指向第 i + 1 层的最左侧结点，
#      赋值给 left_most 后，继续处理第 i + 1 层即可。
# 
#
#      时间复杂度：O(n)
#          1. 需要遍历全部 O(n) 个结点
#      空间复杂度：O(1)
#          1. 只需要维护常数个额外变量即可


"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        # left_most 表示每一层最左侧结点，
        # 初始化为第一层最左侧结点 root
        left_most: Node = root
        # 当前层还有结点时，可以继续处理下一次的 next 指针
        while left_most:
            # 维护一个下一层基于 next 的链表，
            # 为了方便处理，这里使用一个哨兵节点
            next_head_pre: Node = Node()
            # 下一层基于 next 的链表的最后一个结点，
            # 方便使用尾插法插入新结点
            next_last: Node = next_head_pre
            # 当前层的结点
            cur: Node = left_most
            # 如果当前层还有结点，则继续处理
            while cur:
                # 如果存在左子结点，则将其加入到下一层的链表尾部
                if cur.left:
                    next_last.next = cur.left
                    next_last = next_last.next
                # 如果存在左子结点，则将其加入到下一层的链表尾部
                if cur.right:
                    next_last.next = cur.right
                    next_last = next_last.next

                # 继续处理当前层的下一个结点
                cur = cur.next

            # next_head_pre.Next 就是下一层的最左侧结点
            left_most = next_head_pre.next

        return root
